### PR TITLE
Extracted OpenInIDE info bar handling to separate class

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Note: the new class isn't used anywhere yet. It will be used by the show hotspots orchestrator.

### CODE REVIEW

File: src/IssueViz.Security/OpenInIDE/Api/IOpenInIDEStateValidator.cs [hunk 1]
```diff
@@ -20,16 +20,12 @@
...
-using SonarLint.VisualStudio.Core;
 using SonarLint.VisualStudio.Core.Binding;
-using SonarLint.VisualStudio.Core.InfoBar;
 using SonarLint.VisualStudio.Integration;
-using SonarLint.VisualStudio.IssueVisualization.Security.HotspotsList;
 
 namespace SonarLint.VisualStudio.IssueVisualization.Security.OpenInIDE.Api
 {
-    internal interface IOpenInIDEStateValidator : IDisposable
+    internal interface IOpenInIDEStateValidator
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/OpenInIDE/Api/IOpenInIDEStateValidator.cs:28]: The changes to the validator class and tests are straightforward - I just removed anything relating to the infobar and output window service.

File: src/IssueViz.Security/OpenInIDE/Api/OpenInIDEFailureInfoBar.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    internal interface IOpenInIDEFailureInfoBar
+    {
+        void Show();
+
+        void Clear();
+    }
+
+    [Export(typeof(IOpenInIDEFailureInfoBar))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class OpenInIDEFailureInfoBar : IOpenInIDEFailureInfoBar, IDisposable
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/OpenInIDE/Api/OpenInIDEFailureInfoBar.cs:39]: The new class is pretty straightforward too. It's almost exactly the same as it was in the validator class. The only difference is the addition of the `Clear` method so callers can remove old bars if necessary (e.g. if we receive a new "open" request that we can process successfully).

File: src/IssueViz.Security.UnitTests/OpenInIDE/Api/OpenInIDEFailureInfoBarTests.cs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.InfoBar;
+using SonarLint.VisualStudio.Integration.UnitTests;
+using SonarLint.VisualStudio.IssueVisualization.Security.HotspotsList;
+using SonarLint.VisualStudio.IssueVisualization.Security.OpenInIDE.Api;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests.OpenInIDE.Api
+{
+    [TestClass]
+    public class OpenInIDEFailureInfoBarTests
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/OpenInIDE/Api/OpenInIDEFailureInfoBarTests.cs:34]: The tests were cut and pasted from the old tests, then tweaked.
The main difference is that the new tests check that the events are registered and unregistered.

File: src/IssueViz.Security.UnitTests/OpenInIDE/Api/OpenInIDEFailureInfoBarTests.cs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            outputWindowService.VerifyNoOtherCalls();
+
+            // Act
+            infoBar.Raise(x => x.ButtonClick += null, EventArgs.Empty);
+
+            outputWindowService.Verify(x => x.Show(), Times.Once);
+            outputWindowService.VerifyNoOtherCalls();
+        }
+
+        private static OpenInIDEFailureInfoBar CreateTestSubjectWithPreviousInfoBar(
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/OpenInIDE/Api/OpenInIDEFailureInfoBarTests.cs:176]: The code in this test appeared multiple times in the old tests so I extracted it to a helper method.

File: src/IssueViz.Security/OpenInIDE/Api/OpenInIDEFailureInfoBar.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            outputWindowService.Show();
+        }
+
+        private void RemoveExistingInfoBar()
+        {
+            if (currentInfoBar != null)
+            {
+                currentInfoBar.ButtonClick -= ShowOutputWindow;
+                currentInfoBar.Closed -= CurrentInfoBar_Closed;
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/OpenInIDE/Api/OpenInIDEFailureInfoBar.cs:83]: One minor change: the previous code only unregistered one of the event handlers.

Comment (rita-g-sonarsource) [reply src/IssueViz.Security/OpenInIDE/Api/OpenInIDEFailureInfoBar.cs:83]: good catch!

**Review by rita-g-sonarsource — APPROVED** [review]

File: src/IssueViz.Security/OpenInIDE/Api/OpenInIDEFailureInfoBar.cs (hunk 5 above)
Comment (duncanp-sonar) [reply src/IssueViz.Security/OpenInIDE/Api/OpenInIDEFailureInfoBar.cs:83]: I must have missed it when I reviewed the code first time, though ;-)

**Pull request merged**